controllers/cluster_controller.go

```diff
@@ -18,44 +18,98 @@ package controllers
-	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
 	client.Client
+	Log    logr.Logger
 	Scheme *runtime.Scheme
+
+	recorder record.EventRecorder
 }
 
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the Cluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	// Fetch the Cluster instance.
+	cluster := &capiv1alpha3.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
 
-	// your logic here
+		return ctrl.Result{}, err
+	}
 
-	return ctrl.Result{}, nil
+	return r.ReconcileDeletion(ctx, cluster, log)
+}
+
+func (r *ClusterReconciler) ReconcileDeletion(ctx context.Context, cluster *capiv1alpha3.Cluster, log logr.Logger) (ctrl.Result, error) {
+	if _, ok := cluster.Annotations[ignoreClusterDeletion]; ok {
+		log.Info(fmt.Sprintf("Found annotation %s. Cluster %s/%s will be ignored for deletion", ignoreClusterDeletion, cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+
+	}
+
+	if deletionApplied(cluster) {
+		log.Info(fmt.Sprintf("Deletion for cluster %s/%s is already applied", cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+	}
+
+	// immediately delete the cluster if defaultTTL has passed
+	if deletionTimeReached(cluster) {
+		err := r.Client.Delete(ctx, cluster, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if err != nil {
+			log.Error(err, fmt.Sprintf("unable to delete cluster %s/%s", cluster.Namespace, cluster.Name))
+		}
+		log.Info(fmt.Sprintf("Cluster %s/%s will be deleted", cluster.Namespace, cluster.Name))
+
+		return ctrl.Result{}, nil
+	}
+
+	// only send marked for deletion event if we still have ~1h before the cluster gets deleted
+	if deletionEventTimeReached(cluster) {
+		log.Info(fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
+		r.clusterDeletionEvent(cluster, fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
```

Comment:
Not a big deal, but if the operator is restarted or runs the reconciliation for whatever reason, the notification will be sent over and over again, right?

Reply:
If we would restart the operator within this time window, yes we would send the event again. But only once, you would need to restart the operator over and over again within the time window (1h) before the cluster gets deleted.

We could fetch the events on that cluster and see if it was already sent. We have the same issue in upgrade-schedule-operator.

Reply:
yeah as I said not a big deal at all. Even if a little bit of noise gets sent to ourselves nobody's gonna die for sure